erizo_controller/erizoJS/models/Connection.js

```diff
@@ -306,69 +280,143 @@ class Connection extends events.EventEmitter {
   removeMediaStream(id, sendOffer = true) {
     const promise = Promise.resolve();
     if (this.mediaStreams.get(id) !== undefined) {
-      const label = this.mediaStreams.get(id).label;
       const removePromise = this.wrtc.removeMediaStream(id);
       const closePromise = this.mediaStreams.get(id).close();
       this.mediaStreams.delete(id);
-      const retryPromise = Helpers.retryWithPromise(
-        this._resendLastAnswer.bind(this, CONN_SDP, id, label, sendOffer, true),
-        RESEND_LAST_ANSWER_RETRY_TIMEOUT, RESEND_LAST_ANSWER_MAX_RETRIES);
-      return Promise.all([removePromise, closePromise, retryPromise]);
+      return Promise.all([removePromise, closePromise]).then(() => {
+        if (sendOffer) {
+          return this.sendOffer();
+        }
+        return Promise.resolve();
+      });
     }
     log.error(`message: Trying to remove mediaStream not found, id: ${id}`);
     return promise;
   }
 
   setRemoteDescription(sdp) {
-    this.remoteDescription = new SessionDescription(sdp, this.mediaConfiguration);
+    const sdpInfo = SemanticSdp.SDPInfo.processString(sdp);
+    this.remoteDescription = new SessionDescription(sdpInfo, this.mediaConfiguration);
+    this._logSdp('setRemoteDescription');
     return this.wrtc.setRemoteDescription(this.remoteDescription.connectionDescription);
   }
 
-  processOffer(sdp) {
-    const sdpInfo = SemanticSdp.SDPInfo.processString(sdp);
-    return this.setRemoteDescription(sdpInfo);
+  addRemoteCandidate(candidate) {
+    this.wrtc.addRemoteCandidate(candidate.sdpMid, candidate.sdpMLineIndex, candidate.candidate);
   }
 
-  processAnswer(sdp) {
-    const sdpInfo = SemanticSdp.SDPInfo.processString(sdp);
-    return this.setRemoteDescription(sdpInfo);
+  onSignalingMessage(msg) {
+    this._logSdp('onSignalingMessage, type:', msg.type);
+    if (msg.type === 'offer' || msg.type === 'answer-dropped') {
+      this._lockNegotiation('processOffer');
+      return this._onSignalingMessage(msg).then(() => {
+        this._unlockNegotiation();
+      });
+    }
+    if (msg.type === 'answer') {
+      if (!this.isNegotiationLocked) {
+        log.warn('message: Received answer and negotiation was not locked, connectionId: ', this.id);
+      }
+      const promise = this._onSignalingMessage(msg);
+      this._unlockNegotiation();
+      return promise;
+    }
+    if (msg.type === 'offer-dropped') {
+      this.isNegotiationLocked = false;
```

Comment:
If there are cases where we only want to unlock the negotiation without executing the next step, I think we could do it explicitly by having two functions `_unlockNegotiation()` and `_executeNextStep()` just like in the client side

Reply:
Also, are you sure we don't want to enqueue `offer-drop`?